### PR TITLE
Added finger counter

### LINKED ISSUE #817: Finger Counter using Hand Tracking
# Aim

- One will get to know that how to count the fingers using a python script.

# Details

- Initially, I will go through hand tracking and then I will use the hand landmarks to count of the fingers. 
- All of this will be happening in realtime and requires close to no installations and configurations.
- I will use OpenCV, along with some built-in modules of python like `time` and `os`.

### PR DESCRIPTION
**Pull request by mehabhalodiya**
# Description

- Initially, I had gone through hand tracking and then I had used the hand landmarks to count of the fingers.
- All of this is happening in realtime and requires close to no installations and configurations.
- I have used `OpenCV`, along with some built-in modules of python like `time` and `os`.

## Fixes #817 
Finger counter using hand tracking

## Have you read the [Contributing Guidelines on Pull Requests](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/CONTRIBUTING.md)?

- [x] Yes
- [ ] No

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines(Clean Code) of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have created a helpful and easy to understand `README.md`
- [x] My documentation follows [`Template for README.md`](https://github.com/avinashkranjan/Amazing-Python-Scripts/blob/master/Template%20for%20README.md)
- [x] My changes generate no new warnings
- [ ] I have added tests/screenshots(if any) that prove my fix is effective or that my feature works.

## Project Metadata

`If there is no-file/nothing to fill the below fields with, then type: none`

`Example`: If no requirements.txt needed/present then type in `Requirments` none => `Requirments: none`

Category:

 - [ ] Calculators
 - [ ] AI/ML
 - [ ] Scrappers
 - [ ] Social_Media
 - [ ] PDF
 - [ ] Image_Processing
 - [ ] Video_Processing
 - [ ] Games
 - [ ] Networking
 - [ ] OS_Utilities
 - [ ] Automation
 - [ ] Cryptography
 - [x] Computer_Vision
 - [ ] Fun
 - [ ] Others
 
Title: Finger Counter using hand tracking

Folder: Finger-Counter

Requirements: requirements.txt

Script: finger_counter.py

Arguments: none

Contributor: mehabhalodiya 

Description: A system that detects a human hand, segments the hand, counts the number of fingers being held up, and displays the finger count, from live video input.

### CODE REVIEW

**Comment by mehabhalodiya** [comment]
I need some help like confirmation if it is working as I got a sudden problem like stuck somewhere and messed up all!

Can you please check from your side? @kaustubhgupta @santushtisharma10 
Kindly help me.

Let me know if it's working well or else suggest me changes. I will go through that.
Afterwards I will create `README.md` file.

File: Finger-Counter/finger_counter.py
Comment (kaustubhgupta) [inline Finger-Counter/finger_counter.py:12]:
```suggestion
folderPath = "./Finger-Counter/FingerImages"
```

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]
Minor change, else it worked at my end

**Comment by mehabhalodiya** [comment]
I will create a README.md file for this by tomorrow.

**Review by kaustubhgupta — COMMENTED** [review]
Create the readme for this

**Comment by mehabhalodiya** [comment]
@kaustubhgupta @santushtisharma10 @antrikshmisri Kindly review this.

File: Finger-Counter/README.md
Comment (kaustubhgupta) [inline Finger-Counter/README.md:13]:
```suggestion
## Hand tracking in real time
```

File: Finger-Counter/README.md
Comment (kaustubhgupta) [inline Finger-Counter/README.md:54]:
```suggestion
## Finger counter
```

**Review by kaustubhgupta — CHANGES_REQUESTED** [review]

**Comment by antrikshmisri** [comment]
@mehabhalodiya This code is exactly the same as [this](https://stackoverflow.com/questions/67010037/error-with-cv2-cvtcolor-shows-that-src-is-empty) code on stackoverflow, unless it was asked by you. @kaustubhgupta @santushtisharma10

**Comment by mehabhalodiya** [comment]
> @mehabhalodiya This code is exactly the same as [this](https://stackoverflow.com/questions/67010037/error-with-cv2-cvtcolor-shows-that-src-is-empty) code on stackoverflow, unless it was asked by you. @kaustubhgupta @santushtisharma10

Hello @antrikshmisri , I wasn't also aware of which you have provided the link like it exists there.
Actually I have gone through all steps and so understood it well that how it works! Even if you feel this wrong then you can suggest me changes.

I understand your point. Since my Code looks too similar, you can suggest/tell for some changes. So that you can verify whether I understand things fully, by refactoring a few parts.

This is my sincere request as I haven't copied from there.
Hope you will understand.

**Comment by kaustubhgupta** [comment]
@antrikshmisri yes the code is way too similar. I think we should give @mehabhalodiya one more chance to improve upon the code as she has a good history of contributions (9 PRs merged: https://github.com/avinashkranjan/Amazing-Python-Scripts/pulls?q=is%3Apr+author%3Amehabhalodiya+is%3Aclosed) 

@santushtisharma10 @antrikshmisri should we give her the chance?

**Comment by mehabhalodiya** [comment]
You all mentors can suggest me changes in this script, I will do that as per your instructions.
Asking to all three of you for some guidance in this so that I can perform it as you want me to do.
Kindly requesting to suggest me something in this.

**Comment by santushtisharma10** [comment]
I agree, we should give her chance

**Comment by antrikshmisri** [comment]
@kaustubhgupta Sure, we can her another chance based on her contributions. @mehabhalodiya Either restructure your code entirely or you can come up with some new features for it.

**Comment by mehabhalodiya** [comment]
> @kaustubhgupta Sure, we can her another chance based on her contributions. @mehabhalodiya Either restructure your code entirely or you can come up with some new features for it.

Yes, thank you for understanding me.
But I am stuck now like what should be modified?
Any of you can suggest me some changes as I am not knowing what to do further?
I need some help like @kaustubhgupta helped me in `Youtube Video Downloader` script.
Waiting for your response.

**Comment by mehabhalodiya** [comment]
Kindly let me know if you want any changes because I myself don't know.
So asking to you all.

**Comment by mehabhalodiya** [comment]
If you feel that it's good to go then please inform me.
If you are not satisfied with this then I will work until and unless you will guide me.
Awaiting...

**Comment by kaustubhgupta** [comment]
Hi @mehabhalodiya,
The hand tracking module can't be changed that much. For the finger counting module, wrap your code in a class (as done in handtracking)

**Comment by mehabhalodiya** [comment]
Hello @kaustubhgupta ,
I have tried doing that but as you can see that we have already imported hand tracking module in finger counter script and have used like: 
```
img = detector.findHands(img)
lmList = detector.findPosition(img, draw = False)
```
and many more like above!

And now if we are using that class concept then you can imagine how it will get messed up as we will be giving new names for variables, methods,... Further, it will be difficult to understand that how it works (i.e. how it is being executed). Call goes from where to where and the flow of code too got mixed as already we are using class in module and now if we are doing same, then the functions which is being called earlier will now get complex. This is my suggestion! 
Still you don't think so then tell me. 
I think you got my point.

Hope you think and find it useful/efficient.

**Comment by kaustubhgupta** [comment]
> Hello @kaustubhgupta ,
> I have tried doing that but as you can see that we have already imported hand tracking module in finger counter script and have used like:
> 
> ```
> img = detector.findHands(img)
> lmList = detector.findPosition(img, draw = False)
> ```
> 
> and many more like above!
> 
> And now if we are using that class concept then you can imagine how it will get messed up as we will be giving new names for variables, methods,... Further, it will be difficult to understand that how it works (i.e. how it is being executed). Call goes from where to where and the flow of code too got mixed as already we are using class in module and now if we are doing same, then the functions which is being called earlier will now get complex. This is my suggestion!
> Still you don't think so then tell me.
> I think you got my point.
> 
> Hope you think and find it useful/efficient.

It won't get any complicated. You can define wCam, hCam, folderpath, previousTime, tipIds as class variables, wrap image reading in a class function  named `image_read` and the whole while loop in a `process` function. Then you can simply create the object of this class and call them in the dunder init  function (if __name__).

**Comment by mehabhalodiya** [comment]
I have done the changes but as I have told you that I am having some issues to execute this script!
Before committing, I want to let you see. So how can I?
I want to know if it's good or there's any mistake at some point and so want to discuss.
Can you tell me what should I do?

**Comment by kaustubhgupta** [comment]
> I have done the changes but as I have told you that I am having some issues to execute this script!
> Before committing, I want to let you see. So how can I?
> I want to know if it's good or there's any mistake at some point and so want to discuss.
> Can you tell me what should I do?

mark this pull request as a draft, push your code and we will check it at our end

**Comment by kaustubhgupta** [comment]
> I have done the changes but as I have told you that I am having some issues to execute this script!
> Before committing, I want to let you see. So how can I?
> I want to know if it's good or there's any mistake at some point and so want to discuss.
> Can you tell me what should I do?

What issues you are facing?

**Comment by mehabhalodiya** [comment]
Let me know if modified changes works or not as I am not able to run this script properly in my laptop. So requesting you to check from your side.

Please!

**Comment by kaustubhgupta** [comment]
> Let me know if modified changes works or not as I am not able to run this script properly in my laptop. So requesting you to check from your side.
> 
> Please!

That's what I am asking why you're not able to run the script? what error you are getting?

**Comment by mehabhalodiya** [comment]
err1

**Comment by kaustubhgupta** [comment]
@antrikshmisri @santushtisharma10 can you check the working?

**Comment by antrikshmisri** [comment]
> @antrikshmisri @santushtisharma10 can you check the working?
> 

Sure, I'll test the script

**Comment by mehabhalodiya** [comment]
Want me to make any changes?
Kindly let me know.

**Comment by kaustubhgupta** [comment]
@antrikshmisri Please update the status for this script so that we can clear it out

**Comment by santushtisharma10** [comment]
image
@mehabhalodiya fix this error

**Comment by mehabhalodiya** [comment]
err

Now?
I have modified last lines.

**Comment by santushtisharma10** [comment]
ok let me check

**Comment by santushtisharma10** [comment]
@mehabhalodiya but there is no output

**Comment by mehabhalodiya** [comment]
> @mehabhalodiya but there is no output

Yes, I am not able to figure out whats's wrong!
Can you help me like by writing what it will show?

**Comment by kaustubhgupta** [comment]
@antrikshmisri @santushtisharma10 Can we have final word on this?

**Comment by mehabhalodiya** [comment]
I think it will be good to go with the earlier script as you have reviewed it and no errors were generated.
This is just a suggestion from my side else whatever you want me to do, I will !

**Comment by antrikshmisri** [comment]
> @antrikshmisri @santushtisharma10 Can we have the final word on this?

Sorry I got tied up a bit so couldn't test this script, I'll test it by today and let you know.

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [inline Finger-Counter/finger_counter.py:15]:
```suggestion
        self.tipIds = [4, 8, 12, 16, 20]
        self.overlayList = []
        self.image_read(self)
        self.process(self, self)
```

**Review by santushtisharma10 — CHANGES_REQUESTED** [review]
@mehabhalodiya do the following changes and let me know if it worked or not

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [inline Finger-Counter/finger_counter.py:13]:
```suggestion
        self.previousTime = 0
        #self.currentTime = time.time()
        # 4 for thumb, 8 for index, 12 for middle, 16 for ring, 20 for pinky finger
```

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [inline Finger-Counter/finger_counter.py:24]:
```suggestion

        for imPath in myList:
            image = cv2.imread(f'{self.folderPath}/{imPath}')
            self.overlayList.append(image)

        print(len(self.overlayList))
```

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [inline Finger-Counter/finger_counter.py:51]:
```suggestion
                # Thumb
                if lmList[self.tipIds[0]][1] > lmList[self.tipIds[0] - 1][2]:
                    fingers.append(1)
                else:
                    fingers.append(0)

                # Four fingers
                for id in range(1,5):
                    if lmList[self.tipIds[id]][2] < lmList[self.tipIds[id] - 2][2]:
                        fingers.append(1)
                    else:
                        fingers.append(0)
```

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [inline Finger-Counter/finger_counter.py:64]:
```suggestion

                h, w, c = self.overlayList[totalFingers - 1].shape
                img[0:h, 0:w] = self.overlayList[totalFingers - 1]

                cv2.rectangle(img, (20, 225), (170, 425), (0, 255, 0), cv2.FILLED)
                cv2.putText(img, str(totalFingers), (45, 375), cv2.FONT_HERSHEY_PLAIN, 10, (255, 0, 0), 25)
            
            self.currentTime = time.time()
            fps = 1/(self.currentTime - self.previousTime)
            self.previousTime = self.currentTime

```

**Review by santushtisharma10 — CHANGES_REQUESTED** [review]

**Comment by mehabhalodiya** [comment]
> @mehabhalodiya do the following changes and let me know if it worked or not

I have done the suggested changes.

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [inline Finger-Counter/finger_counter.py:11]:
```suggestion
        self.hCam = 480
        self.folderPath = "./Finger-Counter/FingerImages"
        self.previousTime = 0
```

**Review by santushtisharma10 — CHANGES_REQUESTED** [review]
change the folder path as well

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [inline Finger-Counter/finger_counter.py:70]:
```suggestion
            cv2.imshow("Image", img)
            if cv2.waitKey(1) == ord('q') : 
                break
                
cv2.destroyAllWindows()
```

**Review by santushtisharma10 — CHANGES_REQUESTED** [review]

File: Finger-Counter/finger_counter.py
Comment (santushtisharma10) [reply Finger-Counter/finger_counter.py:70]: @mehabhalodiya the script is working on my local system, just add these final changes.

Comment (mehabhalodiya) [reply Finger-Counter/finger_counter.py]: Thank you for helping me wherever I got stuck :)

**Review by santushtisharma10 — APPROVED** [review]

**Review by kaustubhgupta — APPROVED** [review]

**Comment by santushtisharma10** [comment]
@antrikshmisri

**Review by antrikshmisri — APPROVED** [review]

**Comment by kaustubhgupta** [comment]
Update the PR template too

**Comment by mehabhalodiya** [comment]
> Update the PR template too

Updated!

**Pull request merged**
